stab_controller_multi_lqr_node.cpp

```diff
@@ -0,0 +1,416 @@
+/   | / ___// ____/ ____/ | / /_  __/ __  |
+/ /| | \__ \/ /   / __/ /  |/ / / / / / / /
+/ ___ |___/ / /___/ /___/ /|  / / / / /_/ /
+/_/  |_/____/\____/_____/_/ |_/ /_/  \____/
+Author: Lionel Gulich <lgulich@student.ethz.ch> - February 2018
+--------------------------------------------------------------------------------
+Code provided under the GNU General Public Licence GNU GPLv3
+--------------------------------------------------------------------------------
+This code provides the framework for the stabilisation control of the sara
+prototype. It uses multiple lqr controllers derived from different equilibrium
+positions of the system. Between the positions the controllers are interpolated.
+For read in of the k matrix array the matrices should be stored in "/resources"
+as a .csv file and named as "k_matrix_0", "k_matrix_1", ...
+For read in of the imu_offset scaler array the vectors should be stored
+in "/resources" as a .csv file and named as "imu_offset_0", "imu_offset_1", ...
+--------------------------------------------------------------------------------
+TODO: is double necessary in matrices or would float be enough?
+--------------------------------------------------------------------------------
+*/
+
+//Include header, ros and other directories
+#include <ros/ros.h>
+#include "stab_controller_multi_lqr_node.h"
+
+//Set namespace
+namespace ascento {
+
+  //Constructor / Destructor
+  StabControllerMultiLqrNode::StabControllerMultiLqrNode(ros::NodeHandle& nh) : maxon_left(1,"USB1", "EPOS4"), maxon_right(2,"USB0", "EPOS4"), nh(nh) {
+    ROS_INFO("stab_controller_multi_lqr node has started.");
+
+    //Initialise class members
+    Initialise();
+
+    //Initialise Threads for reading position/velocity and writing torques to maxon motors
+    std::thread ReadLeft(&StabControllerMultiLqrNode::ReadLeft,this);
+    std::thread ReadRight(&StabControllerMultiLqrNode::ReadRight,this);
+    std::thread WriteLeft(&StabControllerMultiLqrNode::WriteLeft,this);
+    std::thread WriteRight(&StabControllerMultiLqrNode::WriteRight,this);
+
+    // Start rated loop
+    RateController();
+
+  }
+  StabControllerMultiLqrNode::~StabControllerMultiLqrNode() {
+    maxon_left.MoveWithTorque(0);
+    maxon_right.MoveWithTorque(0);
+  }
+
+  //Setup methods
+  void StabControllerMultiLqrNode::ResetStates() {
+    encoder_offset_left = maxon_left.GetPosition()*2*M_PI/encoder_quadcounts;
+    encoder_offset_right = maxon_right.GetPosition()*2*M_PI/encoder_quadcounts;
+  }
+  void StabControllerMultiLqrNode::Initialise() {
+    // Read parameters from server
+    if (!ReadParameters()) {
+      ROS_ERROR("Could not read parameters.");
+      ros::requestShutdown();
+    }
+
+    InitSubPub();
+    InitSrv();
+
+    // Setup Controller
+    ResetStates();
+    Eigen::VectorXd u_sat = Eigen::MatrixXd(NO_OF_INPUTS,1);
+    u_sat << saturation_ec90;
+    controller.SetUSat(u_sat);
+    input_delays_left.setZero(NO_OF_DELAYS, 1);
```

Comment:
Again for eigen vectors for only need to give 1d inputs so you don't need the ", 1" on the end